ARCHITECTURE.md

```diff
@@ -0,0 +1,352 @@
+# smol-rs Architecture
+
+The architecture of [`smol-rs`].
+
+This document describes the architecture of [`smol-rs`] and its crates on a high
+level. It is intended for new contributors who want to quickly familiarize
+themselves with [`smol`]'s composition before contributing. However it may also
+be useful for evaluating [`smol`] in comparison with other runtimes.
+
+## Thousand-Mile View
+
+[`smol`] is a small, safe and fast concurrent runtime built in pure Rust. Its
+primary goal is to enable M:N concurrency in Rust programs; multiple coroutines
+can be multiplexed onto a single thread, allowing a server to handle hundreds
+of thousands (if not millions) of clients at a time. [`smol`] is intended to
+work on any scale; [`smol`] should work for programs with two coroutines as well
+as two million.
+
+On an architectural level, [`smol`] prioritizes maintainable code and clarity.
+[`smol`] aims to provide the performance of a modern `async` runtime while still
+remaining hackable. This philosophy informs much of the decisions in [`smol`]'s
+codebase and differentiate it from other contemporary runtimes.
+
+On a technical level, [`smol`] is a work-stealing executor built around a
+one-shot asynchronous event loop. It also contains a thread pool for filesystem
+operations and a reactor for waiting for child processes to finish. [`smol`]
+itself is a meta-crate that combines the features of numerous subcrates into a
+single `async` runtime-based package.
+
+The division is as follows:
```

Comment:
Somehow this sounds like a very German way of putting it. :) How about "smol-rs consists of the following crates:"